python/ray/data/datasource/file_based_datasource.py

```diff
@@ -101,6 +101,7 @@ def __call__(
         )
 
 
+@DeveloperAPI
```

Comment:
This abstraction is used as an argument in the public `ds.write_*()` [APIs](https://github.com/ray-project/ray/blob/52b1e4b6722ff681f536e8df3a12754b6018e702/python/ray/data/dataset.py#L1858); should this also be considered a public API?

Reply:
Shall we move it to impl then?

Or take one step back: can this parameter be removed? I may miss the intention here, but it seems DefaultBlockWritePathProvider is only implementation as of now. The logic is simple enough that it can be internalized by those write_*() APIs.